### ISSUE TITLE
Tensor.sort()[:topk] is faster than Tensor.topk() on cuda

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#22812]
## To Reproduce
```python
import time
import torch

data = torch.rand(2000000, dtype=torch.float32, device=torch.device('cuda:0'))
num_topk = 1000

def topk1():
    return data.topk(num_topk, sorted=False)

def topk2():
    sort, idx = data.sort(descending=True)
    return sort[:num_topk], idx[:num_topk]

def benchmark(f, iter, warmup):
    for k in range(warmup): f()
    start = time.perf_counter()
    for k in range(iter): f()
    torch.cuda.synchronize()
    return time.perf_counter() - start

print(benchmark(topk1, 100, 3))
print(benchmark(topk2, 100, 3))
print(benchmark(topk1, 100, 3))
print(benchmark(topk2, 100, 3))
```
On a P100, it prints:
```
2.557645708322525
0.3197173401713371
2.5575470123440027
0.31977611407637596
```

-- my own implementation of `topk2` is faster than the pytorch `.topk()`

## Expected behavior

`topk1` needs to be at least as fast as `topk2`.

## Environment

Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 410.79
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.4.3
[pip3] numpy==1.16.4
[pip3] numpydoc==0.7.0
[pip3] torch==1.1.0
[pip3] torchvision==0.2.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] mkl-service               1.1.2            py36he904b0f_5  
[conda] mkl_fft                   1.0.6            py36hd81dba3_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] torch                     1.1.0                    pypi_0    pypi

cc @ngimel @VitalyFedyunin @fmassa

**Comment by j-min** [comment]
Any updates? This is still reproducible in torch 1.5.1

**Comment by qinjian623** [comment]
Still reproducible in torch 1.6 on NV T4 GPU.

But the performance depends on the shape of data.

If you set data to 2D, and invoke topk while setting dim=1, topk1 will be 10x faster than topk2.

Here is my case:
```
0.0060068462044000626
0.07301453873515129
0.005997709929943085
0.06130932830274105
```

I think this performance gap comes from the different kernel lunch strategy between topk and sort.

**Comment by ngimel** [comment]
Yes, for 1d tensors this result is expected. Is it a common enough usecase so that pytorch should have heuristics for this?

**Comment by ppwwyyxx** [comment]
I found that for 2d tensors with shape AxB and `topk(dim=1)`, the same issue exists when A is very small (<= 4) but B is large.
This is common among object detection models since they often involve topk operation and often use small batchsize (and batchsize 1 in inference).

**Comment by heilaw** [comment]
I am also seeing similar issues on a RTX2080 Ti for 2d tensors with small A and large B. @ngimel Can you explain more on why this behavior is expected on 1d tensors or 2d tensors with small A and large B?

**Comment by ngimel** [comment]
Should be fixed now, please reopen if not.

**Comment by elephaint** [comment]
Issue is still present in my setup with PyTorch 1.10.0+Cuda11.3 on Windows 11, on an RTX2080Ti. Timings when I run the code from @ppwwyyxx:

```
1.4280233999999155
0.10270969999999124
1.3829441999999972
0.1040255000000343
```

So topk is still about 10x slower.

**Comment by ngimel** [comment]
It was fixed after 1.10 was released

**Comment by elephaint** [comment]
@ngimel Thanks, that was not clear to me. From which version is it fixed? Or should I just wait for 1.11? And do you know how I can check going-forward in which version the fix will be included? (to prevent my question....)

**Comment by ngimel** [comment]
You can get nightly build where the fix is included, and it will also be included in 1.11. Usually the issues are closed when they are fixed on master (and thus will appear in nightly builds soon).

**Comment by elephaint** [comment]
Ah, great, thanks!

**Comment by leo-liuzy** [comment]
@ngimel I checked it; there's still huge gaps between topk and sorted[:topk].
My setting: torch-1.10.2+cu113 torchaudio-0.10.2+cu113 torchvision-0.11.3+cu113
GPU: v100 32gb
I got
```
2.3422073060646653
0.08344848733395338
2.3134637447074056
0.08439041022211313
```

**Comment by ringohoffman** [comment]
This should be reopened. On `torch==2.1.1+cu121` on NVIDIA RTX A6000:

```python
import torch
t = torch.randn(1, 3, 224, 224, device="cuda:0")
```

```python
%%timeit
half_k = t.numel() // 2
with torch.no_grad():
    t.view(-1).sort(descending=True).values[half_k]
# 64.5 µs ± 19.2 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

```python
%%timeit
half_k = t.numel() // 2
with torch.no_grad():
    t.view(-1).topk(half_k).values[-1]
# 136 µs ± 1.05 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

```python
%%timeit
half_k = t.numel() // 2
with torch.no_grad():
    t.view(-1).kthvalue(half_k)
# 455 µs ± 1.46 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

`Tensor.sort()` is still 2.1x faster on this example than `Tensor.topk()` and 7.1x faster than `Tensor.kthvalue()`.
